File: backend/app/schemas/stats.py
"""Statistics schemas."""
from typing import Optional
from pydantic import BaseModel
from app.schemas.test import TestSessionResponse, AnswerDetail


class LevelDistribution(BaseModel):
    level: int
    count: int


class RecentTest(BaseModel):
    id: str
    student_id: str
    student_name: str
    student_school: Optional[str] = None
    student_grade: Optional[str] = None
    score: Optional[int] = None
    determined_level: Optional[int] = None
    rank_name: Optional[str] = None
    rank_label: Optional[str] = None
    total_questions: int = 0
    correct_count: int = 0
    duration_seconds: Optional[int] = None
    completed_at: Optional[str] = None
    test_type: str = "test"  # "test" | "mastery"


class ScoreTrend(BaseModel):
    date: str
    avg_score: float
    count: int


class TestHistoryItem(BaseModel):
    test_date: str
    accuracy: int
    determined_level: Optional[int] = None
    rank_name: Optional[str] = None
    rank_label: Optional[str] = None
    correct_count: int
    total_questions: int
    duration_seconds: Optional[int] = None


class TestHistoryResponse(BaseModel):
    history: list[TestHistoryItem]


class WordStat(BaseModel):
    word_id: str
    english: str
    korean: str
    accuracy: float  # 0-100
    attempt_count: int
    avg_time_seconds: Optional[float] = None


class WordStatsResponse(BaseModel):
    lowest_accuracy: list[WordStat]
    slowest_response: list[WordStat]


class RadarMetrics(BaseModel):
    vocabulary_level: float  # 어휘수준 0-10
    accuracy: float          # 정답률 0-10
    speed: float             # 속도 0-10
    vocabulary_size: float   # 어휘사이즈 0-10


class MetricDetail(BaseModel):
    key: str        # "vocabulary_level" | "accuracy" | "speed" | "vocabulary_size"
    name: str       # "어휘수준" | "정답률" | "속도" | "어휘사이즈"
    my_score: float
    avg_score: float
    description: str
    raw_value: Optional[str] = None


class PeerRanking(BaseModel):
    percentile: int
    total_peers: int


class EnhancedTestReport(BaseModel):
    test_session: TestSessionResponse
    answers: list[AnswerDetail]
    radar_metrics: RadarMetrics
    metric_details: list[MetricDetail]
    peer_ranking: Optional[PeerRanking] = None
    grade_level: str
    vocab_description: str
    recommended_book: str
    total_time_seconds: Optional[int] = None
    category_times: dict[str, int] = {}


class MasteryAnswerDetail(BaseModel):
    question_order: int
    word_english: str
    word_korean: str
    correct_answer: str
    selected_answer: Optional[str] = None
    is_correct: bool
    word_level: int
    time_taken_seconds: Optional[float] = None
    stage: int


class MasteryWordSummary(BaseModel):
    word_id: str
    english: str
    korean: str
    final_stage: int
    total_attempts: int
    correct_count: int
    accuracy: float  # 0-100
    avg_time_sec: Optional[float] = None
    mastered: bool


class MasterySessionData(BaseModel):
    id: str
    student_id: str
    total_questions: int
    correct_count: int
    determined_level: Optional[int] = None
    score: Optional[int] = None
    started_at: Optional[str] = None
    completed_at: Optional[str] = None
    best_combo: int = 0
    words_practiced: int = 0
    words_advanced: int = 0
    words_demoted: int = 0


class MasteryReportResponse(BaseModel):
    session: MasterySessionData
    answers: list[MasteryAnswerDetail]
    radar_metrics: RadarMetrics
    metric_details: list[MetricDetail]
    peer_ranking: Optional[PeerRanking] = None
    grade_level: str
    vocab_description: str
    recommended_book: str
    total_time_seconds: Optional[int] = None
    word_summaries: list[MasteryWordSummary] = []


class DashboardStats(BaseModel):
    total_students: int
    total_words: int
    total_tests: int
    avg_score: float
    avg_time_seconds: float
    level_distribution: list[LevelDistribution]
    recent_tests: list[RecentTest]
    weekly_test_count: int
    today_test_count: int
    score_trend: list[ScoreTrend]

File: frontend/src/services/stats.ts
/**
 * Statistics/analytics API service.
 */
import api from './api';
import type { EnhancedTestReport, MasteryReport } from '../types/report';

export interface DashboardStats {
  total_students: number;
  total_words: number;
  total_tests: number;
  avg_score: number;
  avg_time_seconds: number;
  level_distribution: { level: number; count: number }[];
  recent_tests: {
    id: string;
    student_id: string;
    student_name: string;
    student_school: string | null;
    student_grade: string | null;
    score: number | null;
    determined_level: number | null;
    rank_name: string | null;
    rank_label: string | null;
    total_questions: number;
    correct_count: number;
    duration_seconds: number | null;
    completed_at: string | null;
    test_type: 'test' | 'mastery';
  }[];
  weekly_test_count: number;
  today_test_count: number;
  score_trend: { date: string; avg_score: number; count: number }[];
}

export interface TestHistoryItem {
  id: string | null;
  test_date: string;
  accuracy: number;
  determined_level: number | null;
  rank_name: string | null;
  rank_label: string | null;
  correct_count: number;
  total_questions: number;
  duration_seconds: number | null;
}

export interface TestHistoryResponse {
  history: TestHistoryItem[];
}

export interface TestResultItem {
  id: string;
  student_id: string;
  student_name: string;
  student_school: string | null;
  student_grade: string | null;
  score: number | null;
  determined_level: number | null;
  rank_name: string | null;
  rank_label: string | null;
  total_questions: number;
  correct_count: number;
  duration_seconds: number | null;
  completed_at: string | null;
  test_type: 'test' | 'mastery';
}

export interface AllResultsResponse {
  results: TestResultItem[];
  total: number;
}

export interface WordStat {
  word_id: string;
  english: string;
  korean: string;
  accuracy: number;
  attempt_count: number;
  avg_time_seconds: number | null;
}

export interface WordStatsResponse {
  lowest_accuracy: WordStat[];
  slowest_response: WordStat[];
}

export const statsService = {
  async getAllResults(params?: {
    search?: string;
    test_type?: string;
    skip?: number;
    limit?: number;
  }): Promise<AllResultsResponse> {
    const response = await api.get<AllResultsResponse>('/api/v1/stats/all-results', { params });
    return response.data;
  },

  async getWordStats(period: string = 'all'): Promise<WordStatsResponse> {
    const response = await api.get<WordStatsResponse>('/api/v1/stats/word-stats', { params: { period } });
    return response.data;
  },

  async getDashboardStats(period: string = 'all'): Promise<DashboardStats> {
    const response = await api.get<DashboardStats>('/api/v1/stats/dashboard', { params: { period } });
    return response.data;
  },

  async getStudentHistory(studentId: string): Promise<TestHistoryResponse> {
    const response = await api.get<TestHistoryResponse>(`/api/v1/stats/student/${studentId}/history`);
    return response.data;
  },

  async getEnhancedReport(studentId: string, testId: string): Promise<EnhancedTestReport> {
    const response = await api.get<EnhancedTestReport>(
      `/api/v1/stats/student/${studentId}/report/${testId}`,
    );
    return response.data;
  },

  async getMasteryReport(studentId: string, sessionId: string): Promise<MasteryReport> {
    const response = await api.get<MasteryReport>(
      `/api/v1/stats/student/${studentId}/mastery-report/${sessionId}`,
    );
    return response.data;
  },
};

export default statsService;
